### DISCUSSION TITLE
Queries Regarding Code Insertion & Execution In A Blog

### DISCUSSION THREAD

**Discussion by samarzayan** [quarto-dev/quarto-cli#3035] [Q&A] [answered]
I had a few questions:
1. Where does the code included in the `.qmd` files get executed in the rendering and preview stages? Is it on my local machine or on Quarto's cloud servers? 
2. What is a better option: Including a Colab file (`.ipynb` file) in `_quarto.yaml` section or writing blocks of code within a `.qmd file and including that in the `yaml` file?

My situation: I am trying to convert some Deep learning model codes into a blog from the fast.ai course.`

I tried to include code blocks in a `.qmd` file but it is resulting in all kinds of errors. Not sure how things are getting executed.

Also, can I add a `.ipynb` file inside a '.qmd' file?  

Thanks in advance!

**Comment by cscheid** [comment, marked as answer]
`quarto render` and `quarto preview` happen locally on your machine.

> I tried to include code blocks in a .qmd file but it is resulting in all kinds of errors. Not sure how things are getting executed.

That code is likely running on your machine. There's a difference in how we process .ipynb files compared to .qmd files. By default, we _do not execute_ ipynb files, and instead we use the results that will be included in them (since in Jupyter notebooks, you can save the result of an execution in the notebook itself).

For a .qmd file, on the other hand, we have to execute all code cells (since we certainly won't have access to previous results).

To answer your question about what is a better option between .ipynb and .qmd files, it depends on what you're attempting to do. If your notebook file includes results that can only be generated by (for example) a machine with access to a GPU or TPU, then you're probably going to need to use .ipynb files as they are.
